### PR TITLE
FEATURE: add full bot support for GPT 3.5

### PR DESCRIPTION
**Pull request by SamSaffron**
Given latest GPT 3.5 16k which is both better steered and supports functions
we can now support rich bot integration.

Clunky system message based steering is removed and instead we use the
function framework provided by Open AI

Tests still need fixing and system needs a bit more debugging, but early
results are very promising!

### CODE REVIEW

File: lib/modules/ai_bot/bot.rb [hunk 1]
```diff
@@ -3,6 +3,25 @@
...
+        attr_reader :functions
+        attr_reader :current_function
+
+        def initialize
+          @functions = []
+          @current_function = nil
+        end
+
+        def add_function(name)
+          @current_function = { name: name, arguments: +"" }
```
Comment (jjaffeux) [inline lib/modules/ai_bot/bot.rb:16]: What is the `+""` achieving here? I would say nothing given it's an empty string but I might not know something here...

Comment (romanrizzi) [reply lib/modules/ai_bot/bot.rb:16]: He will later mutate the string using `<<`

Comment (jjaffeux) [reply lib/modules/ai_bot/bot.rb:16]: ok I see

```ruby
irb(main):009:0> +"".freeze << "foo"
=> "foo"
irb(main):010:0> "".freeze << "foo"
(irb):10:in `<main>': can't modify frozen String: "" (FrozenError)
	from /Users/joffreyjaffeux/.asdf/installs/ruby/3.2.1/lib/ruby/gems/3.2.0/gems/irb-1.6.3/exe/irb:9:in `<top (required)>'
	from /Users/joffreyjaffeux/.asdf/installs/ruby/3.2.1/bin/irb:25:in `load'
	from /Users/joffreyjaffeux/.asdf/installs/ruby/3.2.1/bin/irb:25:in `<main>'
```

Well first time I see that syntax

File: lib/modules/ai_bot/open_ai_bot.rb [hunk 2]
```diff
@@ -18,7 +18,7 @@ def prompt_limit
         if bot_user.id == DiscourseAi::AiBot::EntryPoint::GPT4_ID
           8192 - 3500
```
Comment (xfalcox) [inline lib/modules/ai_bot/open_ai_bot.rb:19]: There is also GPT-4 32k

Comment (SamSaffron) [reply lib/modules/ai_bot/open_ai_bot.rb:19]: we don't have access yet afaik

Comment (SamSaffron) [reply lib/modules/ai_bot/open_ai_bot.rb:19]: plus MEGA expensive.

**Review by romanrizzi — APPROVED** [review]
Looks good to merge. We should probably do some cleanup/refactoring in the future to make it easier to follow - it has become quite complex.

**Comment by SamSaffron** [comment]
Thanks Roman, yeah I agree, this is getting quite complex and we are going to have to support functions with other LLMs which is going to introduce even more complexities.

**Pull request merged**